docker/docker-init.sh

```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
+
+# start dockerd if env var is set
+if [ "${DOCKERD}" = "1" ]
+then
+    [ -e /var/run/docker.sock ] && rm /var/run/docker.sock
+    dockerd &
+fi
+
+# start ssh agent if socket not mounted from host
+if [ ! -e "${SSH_AUTH_SOCK}" ]
+then
+    eval "$(ssh-agent -s)"
+fi
+
+# install private ssh key if one is provided
+if [ -n "${SSH_PRIVATE_KEY}" ]
+then
+    echo "${SSH_PRIVATE_KEY}" | tr -d '\r' | ssh-add -
+fi
+
+# try to determine if an executable was provided or just args
+if [ "${1}" = "balena" ] || [ -x "${1}" ] || "${1}" -v >/dev/null 2>&1
```

Comment:
This isn't working as intended, I'm retesting it now with

```bash
# try to determine if an executable was provided or just balena args
if ! balena "${1}" --help >/dev/null 2>&1
then
    exec "$@"
else
    exec balena "$@"
fi
```

Reply:
> if ! balena "${1}" --help >/dev/null 2>&1

hmm a disadvantage of this approach is that every CLI command results in invoking the CLI twice, which causes additional "unnecessary" delay and lag in CLI usage. I think the `--help` form also tracks usage with Amplitude.com, and sometimes this can add variable delays. Oh, at least you could disable this tracking with:

```
if ! BALENARC_NO_ANALYTICS=1 balena "${1}" --help >/dev/null 2>&1
```

But even disabling analytics, I am not fan of this approach of executing the CLI twice, because it is slow.

Reply:
Okay, let me think about it some more.

Reply:
The catch so far has been some balena commands are also valid executables in the system path, so just saying "is this cmd an executable file in PATH" will fall over with things like `ssh` and some others.

I'm trying to be too clever by supporting both args and commands in the CMD field, but I'm determined to make it work in a predictable way!

I was pretty close when I didn't check `[ -x "$(which "${1}")" ]` and only checked `[ -x "${1}" ]` but then it fails when providing `bash` instead of `/bin/bash`. I might go back to this as it worked the best.

Is there a file in the source code where we could quickly grep whether something is a real balena command?

Reply:
Related, I'm going to limit the cases where the ssh-agent starts to only when the SSH_PRIVATE_KEY is provided and the agent is not already mapped from the host.

Reply:
> Is there a file in the source code where we could quickly grep whether something is a real balena command?

Yes, we do actually: `oclif.manifest.json`
```
{ ...
  "commands": {
    "apps": {
    ...
    "build": {
    ...
    "deploy": {
    ...
```
It's not committed to the git repo, but produced by `npm run build:fast`. We could extract the list of command names during the image build (in the Dockerfile), to even reduce JSON parsing time at runtime (it's a large file, 123KB, as it contains all the flags as well...)

Reply:
That `oclif.manifest.json` file is handy, but this is all getting a bit messy. What do you think about these three options:

### 1. Try to determine if command is executable or args

```bash
if [ "${1}" = "balena" ] || [ -x "${1}" ]
then
    exec "$@"
else
    exec balena "$@"
fi
```

- shell can be started with `/bin/bash`
- the default CMD could be `/bin/bash` if we assume a container shell is the common use case?
- the default CMD could be `--help` if we assume a single balena command is the common use case?
- failing to provide the full path to an exec will treat it as an arg (eg. `bash` => `balena bash`)

### 2. Assume command is executable

```bash
exec "$@"
```

- shell can be started with `bash`
- one-liner docker run commands require `balena ` prefix
- the default CMD could be `bash` if we assume a container shell is the common use case?
- the default CMD could be `balena --help` if we assume a single balena command is the common use case?

### 3. Assume command is args

```bash
exec "balena $@"
```

- **shell cannot be started without skipping dockerd and ssh-agent init steps via custom entrypoint**
- one-liner docker run commands do not require `balena ` prefix
- the default CMD could be `--help` if we assume a single balena command is the common use case?

Reply:
I've added a commit that attempts to deliver "option zero", the one you described as getting a bit messy. :-)  Do you think it works? I am happy for you to modify it!

Reply:
It's like option 1, but better discriminates when the argument is something like `bash` or `whoami`, treating the latter as a CLI command but not the former, giving priority to CLI commands when there is a clash.